lib/spack/spack/config.py

```diff
@@ -432,7 +432,7 @@ def get_config_filename(scope, section):
     return scope.get_section_filename(section)
 
 
-def update_config(section, update_data, scope=None):
+def update_config(section, update_data, scope=None, overwrite=False):
```

Comment:
I'm finding this change a bit confusing given the docstring for the method. For the non-list case I think you get the desired behavior if you remove all existing entries from the configuration:

```
    else:
        configuration.clear()
        configuration.update(update_data)
```

I think it would also be reasonable to replace the function body with:

```
    validate_section_name(section)  # validate section name
    scope = validate_scope(scope)  # get ConfigScope object from string.

    # read only the requested section's data.
    scope.sections[section] = {section: update_data}
    scope.write_section(section)
```

The only issue would be if non-list entities which support the `.update` function can't be written out. That being said none of the tests check for this.

Reply:
@alalazo does this seem agreeable?

Reply:
@scheibelp Sure, I'll try to get to this asap.

Reply:
I checked briefly all the places where `update_config` got called, and I agree with what you say above. If I had to be pedantic, probably `update_config` is the wrong name for this function as it reminds of dictionary updates.